text/3579-pki.md

```diff
@@ -0,0 +1,262 @@
+
+## Out of Scope
+
+This RFC does not attempt to solve:
+- Identities or signing of Rust Project members
+- Identities or signing of crate authors
+- Answering open questions around the benefits or capabilities of any next-generation repository signing solutions (e.g. sigstore or the update framework (TUF))
+- Specific design or implementation strategies around binary signing (e.g. for Windows or macOS binaries)
+
+# Motivation
+[motivation]: #motivation
+
+The Rust Project currently lacks any official Public Key Infrastructure (PKI) required to perform proper package, mirror, code, and release signing across the project. There is high demand for capabilities that require such PKI infrastructure, such as mirroring and binary signing. This RFC provides the necessary cryptographic infrastructure to unblock such work, and an initial usage for cryptographic verification of Rust releases and `crates.io` downloads.
+
+We have chosen this specific approach to give us the most flexibility in choosing implementations for cryptographic security in future RFCs and implementations. Multiple problem spaces currently exist which can leverage an internal PKI - the most visible and obvious being crate and release signing. However, other avenues also exist - such as mirror validation or internal and external services communication. We see a future in which this Certificate Authority can be leveraged for meeting the needs of these problems, and consider it the first stepping stone towards a solution.
+
+## Mirroring
+
+There is currently no way for end users to cryptographically guarantee, in isolation, that an index entry, crate package, or Rust release was definitively sourced from the Rust Project. Although the entirety of the current end-to-end system provides some protections (hashes in the index of crate files, and HTTPS for all communications), this precludes many out-of-band mirroring scenarios and is subject to multiple attack vectors.
+
+This proposal adds cryptographic security for multiple mirroring scenarios:
+- Rust releases (downloaded directly or via `rustup`)
+- `crates.io` index
+- `crates.io` packages
+
+Today, for `crates.io` and official infrastructure, we strictly distribute from Rust Project controlled sources (our CDN network). This prevents us from leveraging mirroring offers from third parties, such as academic institutions, to host copies of crates.io packages as well as Rust releases. Having a set of root cryptographic keys gives us the tools to begin exploring multiple methods of securing both internal and external mirroring and propagation problems.
+
+We leave the implementation of these solutions for a future RFC - however, this provides us with the first steps in having the available cryptographic keys to begin exploring and discussing solutions to these problems.
```

Comment:
This is making an assumption (IMO not justified here) that the right way to solve trust problems for mirroring is a PKI hierarchy with long-lived roots that are pinned in clients. That does not feel like an obvious claim to me.

(This is mostly just an example of my point above that this seems like it's reaching ahead for a particular solution and *not* saying why we are rejecting technical solutions that are alternatives. But since this isn't proposing any concrete scheme, it's also hard for me to say "why is this a requirement?" or "non-requirement?"

Reply:
We've pushed changes that expand on why to bridge a CA/PKI, and given more details on the threat model. Beyond that, I think the rest of this will be addressed by posting a subsequent concrete RFC for crates.io index signing/mirroring, which we're working on now.